File: CS50x2023-Practice/Week1-C/Solution/credit.c
#include <stdio.h>
#include "cs50.h"

char check_card(long n);    // check card numbers' company

char check_validity(int s); // check card numbers' validity

int main(void)
{
    int i = 0;
    char card; // Store Card Company
    long number; // Store Card Number
    char validity;
    int index = 1;
    int s1 = 0; // Sum of odd digits
    int s2 = 0; // Sum of even digits*2
    int sum = 0; // Total Sum

    number = get_long("Number: "); // write a program that prompts the user for a credit card number

    while (number != 0)
    {
        i = number % 10;    //  Get the first digit on the right

        if (index % 2 == 1)
        {
            s1 = s1 + i;
        }
        else
        {
            int even = i * 2;
            if (even > 9)
            {
                s2 = s2 + (even / 10) + (even % 10);    // If the even is greater than 9, you need to get each bit separately as a separate number
            }
            else
            {
                s2 = s2 + even;
            }
        }

        if (number >= 10 && number <= 99)  // Start with the first two digits of the card number
        {
            card = check_card(number);
        }
        number = number / 10;
        index++;
    }

    sum = s1 + s2;
    validity = check_validity(sum);
    index = index - 1;
    //  Reports (via printf) whether it is a valid American Express, MasterCard, or Visa card number, per the definitions of each’s format herein
    if (validity == 'V' && card == 'A' && index == 15)  // American Express uses 15-digit numbers
    {
        printf("AMEX\n");
    }
    else if (validity == 'V' && card == 'M' && index == 16) // MasterCard uses 16-digit numbers
    {
        printf("MASTERCARD\n");
    }
    else if (validity == 'V' && card == 'V' && (index == 13 || index == 16)) // Visa uses 13- and 16-digit numbers
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

char check_card(long n)
{
    if (n == 34 || n == 37) //  All American Express numbers start with 34 or 37
    {
        return ('A');   // American
    }
    else if (n >= 51 && n <= 55)    // Most MasterCard numbers start with 51, 52, 53, 54, or 55
    {
        return ('M');   // MasterCard
    }
    else if ((n / 10) == 4) // All Visa numbers start with 4
    {
        return ('V');
    }
    else
    {
        return ('I');   // Invalid
    }
}

char check_validity(int s)
{
    if (s % 10 == 0)    //  if the total modulo 10 is congruent to 0
    {
        return ('V');   // Valid
    }
    else
    {
        return ('I');   // Invalid
    }
}

File: CS50x2023-Practice/Week6-Python/Solution/sentimental-credit/credit.py
# TODO: import get_string from cs50 library
from cs50 import get_int, get_string

while True:
    number = get_int("Number: ")
    if (number >= 0):
        break

tmp = number
sum = 0
i = 0

while number > 0:
    digit = number % 10 # stores the last digit of the number
    if (i % 2 == 0):
        sum += digit
    else:
        if (digit * 2 >= 10):
            sum += digit * 2 // 10
            sum += digit * 2 % 10
        else:
            sum += digit * 2
    number = number // 10
    i += 1

number = tmp

if (sum % 10 == 0): # checks if last number of the total sum is 0
    if (number // 10000000000000 == 34 or number // 10000000000000 == 37):
        print("AMEX")
    elif (number // 100000000000000 >= 51 and number // 100000000000000 <= 55):
        print("MASTERCARD")
    elif (number // 1000000000000 == 4 or number // 1000000000000000 == 4):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")